TypedDataset skeleton

kanterov opened the pull request:
This pull request adds `TypedDataset`. In contrast to `Encoder` and `Dataset` from Spark it uses shapeless to derive encoders, it allows to customize encoding of any Scala type and gives compile-time guarantees.

`TypedDataset` gets syntax from `frameless.syntax`, it leads to better modularity and gives library user a choice which syntax to use. In particular, it might be useful to wrap a side-effecting code into effectful monads like `scalaz.concurrent.Task` and provide support for cats and scalaz, see `EffectfulOps`.

General operations, such as `select`, `filter`, `groupBy` are implemented, however, lots of things are still missing. I feel this approach could be replicated to implement the rest of functionality. I've found very effective to verify functionality with property-based tests, especially it was a surprise for me how `sum` and `avg` behave on different data types. 

I tried not to abuse shapeless in user-facing methods to get nice type-inference from IDEs and good compile times. I think we should provide boilerplate with reasonable arities as well as generic arity alternatives because it helps people that heavily rely on IDE and can't read type signatures with 5+ type classes.

OlivierBlanvillain commented:
Wow looks awesome! I gave this a quick read and I can't find some of the core definitions, such as `TypedEncoder`, `TypedColumn` and `TypedDataset`, did you forgot to commit some files?

kanterov commented:
@OlivierBlanvillain oh god, thanks, I've pushed missing files

OlivierBlanvillain commented:
I still need to spend more time on `ProductEncoder.scala`, `RecordEncoder.scala` and `TypedEncoder.scala`, I'm getting too tired for reading code...

As a general comment, I have to say that I'm not really a fan of having `class TypedDataset[A]` with any empty body, and all the methods implemented via implicit conversions spread over many files. I find really more accessible to have it all in one big fat file like it's done in Sparks source code. I don't know it plays with IDE but as a reader I definitely prefer to have it "all in one place". I think the use of this pattern should be restricted to things such as `EffectfulOps` where we might want to provide several alternative of a single method like `collect`, maybe one in `blocking._`, one in `std._` (for `Future`), and one in `scalaz._` for `Task`.

> I think we should provide boilerplate with reasonable arities as well as generic arity alternatives

I disagre, I would really vote to start with a minimalist code base, and only provide boilerplate variants if users really ask for it.

Regarding syntax, here are a few details that I noted while reviewing:

- Long function signature with implicits: I find [this](https://github.com/milessabin/shapeless-type-class-derivation-2015-demo/blob/master/src/main/scala/derivation/derivation.scala#L182) to be the most readable way to write the, shapeless code base mixes many styles for that but I would love to keep it uniform if that's not asking too much :)

- Space before context bound: I think they help to differentiate them from regualar typing.

- Names of implicits. I think the following convention helps readability:

    - Use meaninful names when the implicit is referenced inside the method body,
    - Use single/two letter otherwise

If you agree with some comments but find them too tendious to fix in this RP I would be happy to leave it as is for now and do in a followup.

OlivierBlanvillain commented:
Dam I think I reviewed on the commit instead of the diff, try not to `push -f` or I think all my comments will be lost :(

kanterov commented:
@OlivierBlanvillain Thanks for the review :+1: I've replied your comments, going to address them. 

> I disagree, I would vote to start with a minimalist code base, and only provide boilerplate variants if users really ask for it.

I implemented this because my team consists of developers of different levels, most of them need this to be productive, so it was kind of requirement from my side.

I will try to optimize implicit parameters, I didn't spend much time on it yet.

kanterov commented:
Regarding `Isomorphism`. The idea is to use this to extend the library with instances for new types like `Usd` and things like Java Time API and JodaTime that are isomorphic to `SQLDate` and `SQLTimestamp`. It would allow using Spark functions like `sum`, `avg`, `weekofyear`, `date_add` on custom types.

I didn't find good encoding for this yet, because we have to build expression tree for Spark, see `IsomorphismTest`. There should be a law to check custom `TypedEncoder` implementations, to verify it you should define an `Isomorphism` with built-in type. Then we can check for evidence of `Isomorphism` in type signatures of `sum`, `avg` and others. Probably there should be something like `IsomophicTypedEncoder`.

kanterov commented:
@OlivierBlanvillain I've just pushed an update.

1. Addressed your comments.
2. Introduced `Job[T]` for methods that start new spark jobs. It has minimal implementation right now.
3. Removed `ops`. Added methods for generic and 1-2-3 arities to `TypedDataset` as a compromise.

I think it is a good starting point because it already took awhile.

adelbertc commented on dataset/src/main/scala/frameless/Job.scala:
```diff
@@ -0,0 +1,31 @@
+
+sealed abstract class Job[T](implicit sc: SparkContext) { self =>
+  /** Runs a new Spark job. */
+  def run: T
```
I wonder if instead we should modify run to be

```scala
def run(sc: SparkContext): T
```

to avoid implicit config passing.

Also as a matter of style, I believe type params starting with `A` tend to be preferred

adelbertc commented on dataset/src/main/scala/frameless/Job.scala:
```diff
@@ -0,0 +1,31 @@
+}
+
+object Job {
+  def apply[A](a: => A)(implicit sc: SparkContext): Job[A] = new Job[A] {
```
Would prefer `apply` to take `SparkContext => A` and rename what you have here to be `const` or `pure` or something

adelbertc commented:
:+1: LGTM

One note: `Job` is basically the Reader monad for Spark jobs, and it's a technique I've heard many people use (myself included) so :+1: 

On a related note, it may be useful to also define `flatMap` and `map` on it for easy composition. It would end up looking something like:

```scala
trait Job[A] {
  def run(sc: SparkContext): A

  def flatMap[B](f: A => Job[B]): Job[B] = ...

  def map[B](f: A => B): Job[B] = ...
}
```

OlivierBlanvillain commented:
LGTM!
